packages/push/src/instruction/exec/when.rs

```diff
@@ -0,0 +1,201 @@
+/// | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
+/// | `true`  | exists | Consumed | Unchanged | ✅ | The code block is executed |
+/// | `false` | exists | Consumed | Consumed | ✅ | The code block is skipped |
+/// | missing | exists | Non-existent | Consumed | ✅ | The code block is skipped |
+/// | exists | missing | Unchanged | Non-existent | ✅ | State is unchanged |
+/// | missing | missing | Non-existent | Non-existent | [❗…](StackError::Underflow) | State is unchanged |
+///
+/// # Errors
+///
+/// If either of the stack accesses returns any error other than a
+/// [`StackError::Underflow`] then this returns that as a [`Error::Fatal`]
+/// error.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct When;
+
+impl NumOpens for When {
+    fn num_opens(&self) -> usize {
+        1
+    }
+}
+
+impl<S> Instruction<S> for When
+where
+    S: Clone + HasStack<PushProgram> + HasStack<bool>,
+{
+    type Error = PushInstructionError;
+
+    fn perform(&self, state: S) -> InstructionResult<S, Self::Error> {
+        let condition = state.stack::<bool>().top();
+        let block = state.stack::<PushProgram>().top();
+        match (condition, block) {
+            // If there is a boolean that is true and a block, discard the boolean
+            // and leave the block so that it may be
+            // performed next.
+            (Ok(true), Ok(_)) => Ok(state).with_stack_discard::<bool>(1),
+            // If there is a boolean that is false and a block, discard both the
+            // boolean and the block since we don't
+            // want to perform that block.
+            (Ok(false), Ok(_)) => Ok(state)
+                .with_stack_discard::<bool>(1)
+                .with_stack_discard::<PushProgram>(1),
+            // If there is no boolean but there is a block, discard the block since
+            // we only want to perform it if there is a
+            // boolean that is true.
+            (Err(StackError::Underflow { .. }), Ok(_)) => {
+                Ok(state).with_stack_discard::<PushProgram>(1)
+            }
+            // If there is a boolean but no block, then we just skip this instruction.
+            (Ok(_), Err(StackError::Underflow { .. })) => Ok(state),
+            // If there is no boolean and no block, then we return an stack underflow
+            // error.
+            (Err(StackError::Underflow { .. }), Err(e @ StackError::Underflow { .. })) => {
+                Err(Error::recoverable(state, e))
+            }
+            // If some other error occurs (e.g., a fatal error), then we just
+            // pass that forward.
+            (Err(e), _) | (_, Err(e)) => Err(Error::fatal(state, e)),
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+mod tests {
+    use crate::{
+        error::into_state::IntoState,
+        instruction::{exec::when::When, ExecInstruction, Instruction, PushInstructionError},
```

Comment:
And here on the other hand we should probably import when through `super::When`, not the full crate path

Reply:
Good catch. And I think I just "broke" that in getting rid of `super::*` everywhere. I'll go fix that.